### PR TITLE
3241 rebuilds cause metadata error and 3627 adding retry logic

### PR DESCRIPTION
**Pull request by AndrewMcClain**
### Intent

Addresses rstudio/rstudio-pro#3241
Addresses rstudio/rstudio-pro#3627

Rerunning a build should not either fail the build or result in mismatched sha's. These updates will attempt to create the file as usual, but monitor response for indications of error, and if we do have an error, attempt to get the file from the repo, grab the original file's sha, and add that to the create payload to create and update payload.

### Approach

in bash and powershell we capture the output of our web request. In bash we check for the root level message element, that indicates a failure.
Powershell swapped from invoke-webrequest to invoke-restmethod so our responses come back to us in json. We have to use a try catch here to catch failures.
After that the proceedure is pretty much the same, get file, get the `sha` field, use that in the `sha` field in the payload and rerun our update.

The nonsensical error messages are because the `sha` field in a Create or Update request indicates an update, and the `sha` provided is the hash of the file in the repo, not the file we're updating, so it was complaining that the sha's didn't match.

### Automated Tests

None, build system only

### QA Notes

Passing builds will test

### Documentation
No docs needed

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: docker/jenkins/publish-build.ps1
Comment (jmcphers) [inline docker/jenkins/publish-build.ps1:124]:
```suggestion
{ "message": "Update $flower build $version in $build", "content": "$base64", "sha": "$updateSha" }
```

File: docker/jenkins/publish-build.sh
Comment (jmcphers) [inline docker/jenkins/publish-build.sh:213]:
```suggestion
   updatePayload="{\"message\":\"Update $flower build $version in $build\",\"content\":\"$base64_contents\",\"sha\":\"$updateSha\"}"
```

File: docker/jenkins/publish-build.sh
Comment (jmcphers) [inline docker/jenkins/publish-build.sh:211]: Do we know for sure that `jq` is installed on all the build containers? If not, do we need to update the Dockerfile to add it?

Comment (AndrewMcClain) [reply docker/jenkins/publish-build.sh:211]: Yeah, jq's included in all the `docker/jenkins` Dockerfiles

Comment (AndrewMcClain) [reply docker/jenkins/publish-build.sh:211]: Actually looks like I'm mistaken, the publish step doesn't happen in the container, it's on the build agent's ami, so I'm not sure

Comment (cm421) [reply docker/jenkins/publish-build.sh:211]: Should we move the publish step into the container? I'm surprised to hear it wasn't already happening in the container

Comment (AndrewMcClain) [reply docker/jenkins/publish-build.sh:211]: It would give us a more controlled, repeatable environment, And I'm not sure what benefit publishing outside the container would give us

Comment (cm421) [reply docker/jenkins/publish-build.sh:211]: I'm working on restructuring the linux jenkinsfile so I'll include that change in there. Graham is doing the windows one so I'll let him know

Comment (AndrewMcClain) [reply docker/jenkins/publish-build.sh:211]: Alright, Then to prevent failing builds, this is blocked until the merge of [#4185](https://github.com/rstudio/rstudio-pro/issues/4185) right?

File: docker/jenkins/publish-build.ps1
Comment (MariaSemple) [inline docker/jenkins/publish-build.ps1:119]: Wouldn't this be a no-op? We get the sha that the dailies page already has and give it back to the dailies page?

File: docker/jenkins/publish-build.sh
Comment (MariaSemple) [inline docker/jenkins/publish-build.sh:183]: Is this because the dailies site calculates the sha itself?

File: docker/jenkins/publish-build.sh
Comment (MariaSemple) [reply docker/jenkins/publish-build.sh:251]: Looks like this wasn't added, so I unresolved it

File: docker/jenkins/publish-build.ps1
Comment (AndrewMcClain) [reply docker/jenkins/publish-build.ps1:119]: We need to retrieve the sha of the file in the dailies page repo to perform an update instead of a create.

Previously we were supplying a the binary sha as the github file sha and in a create, that didn't cause an issue, it was ignored. In the update case the github file sha didn't match the binary sha, producing the confusing error messages.

What this code is doing is getting the github file sha, so we can resubmit our change as an update, instead of as a create. The github api is using the file sha as a way of making sure a commit doesn't interleave and we don't end up stomping changes. See https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents

File: docker/jenkins/publish-build.sh
Comment (AndrewMcClain) [reply docker/jenkins/publish-build.sh:183]: See above

**Review by MariaSemple — APPROVED** [review]
Once we merge in the epic branch (which we should do as soon as CB releases), LGTM!

**Comment by AndrewMcClain** [comment]
Reopend this it looks like the linux retry / update logic was never included.

**Comment by cm421** [comment]
LGTM, thanks for noticing it was missing!

**Review by cm421 — APPROVED** [review]

**Review by MariaSemple — APPROVED** [review]
LGTM!

**Pull request merged**
